src/Connection/CallbackConnectionProvider.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace SMW\Connection;
+
+use Closure;
+
+/**
+ * @license GNU GPL v2+
+ * @since 3.0
+ *
+ * @author mwjames
+ */
+class CallbackConnectionProvider implements ConnectionProvider {
+
+	/**
+	 * @var Closure
+	 */
+	private $callback;
+
+	/**
+	 * @var mixed
+	 */
+	private $connection;
+
+	/**
+	 * @since 3.0
+	 *
+	 * @param Closure $callback
+	 */
+	public function __construct( Closure $callback ) {
+		$this->callback = $callback;
+	}
+
+	/**
+	 * @since 3.0
+	 *
+	 * @return mixed
```

Comment:
Should this not be a connection thing?

Reply:
No, it is a generic callback therefore you type. The invoker is responsible to make the right choice.

Reply:
I understand that this function cannot guarantee the type that is returned. Though I think this does not really matter for the return type. The return type is about indicating what it should return. Of course this should match what the callback is supposed to return. Unless I am misunderstanding the code, in both cases this should be a connection, whatever a connection is. Weirdly enough the ConnectionProvider interface does not mention a specific type. Is that understanding wrong?